FLOIP/flow-builder
# Vmo 690 with runner v1

Update: @bzabos  has done an initial review and is happy for this to move to QA.

**Notes for QA**

When testing, this `builder.config.json.txt` file can be used in the place of the default to make testing easier. Simply copy across the file and remove the `.txt` extension. It contains groups and properties and one additional languag. When copied over you can see what has changed in the diff.

[builder.config.json.txt](https://github.com/FLOIP/flow-builder/files/6421720/builder.config.json.txt)

All tests should be done for both the upload and by pasting the contents of the test files into the Paste field

- Test save and continue is disabled when the import page is in an invalid state
- Test invalid json - e.g. a file containing some random text
  - You should see an error saying you did not use valid json
- Test multiple flows - take a look at `src/store/builder/5b8c87d6-de90-4bc4-8668-4f040000006e-flow-multi.json`
  - You should see an error saying the importer only supports single flows
- Test lang mistmatch - `src/store/builder/5b8c87d6-de90-4bc4-8668-4f040000006e-flow-invalid-lang.json`
  - You should see the lang mismatches highlighted
  - You should be able to resolve by selecting from the dropdown and clicking the button
  - The language in the flow json should be appended to the languages list
  - Any resource using the old language id should be set to use the new language id
  - Edit the test of languages so that the id or other attributes no longer match the language previously matched in builder.config.json
  - You should see a loading spinner and then the edited language should be removed from the flow. Then you should see the matcher to fix the language you broke 
- Test group mismatch - `src/store/builder/5b8c87d6-de90-4bc4-8668-4f040000006e-flow-block-group.json`
  - You should see the group mismatches highlighted
  - You should be able to resolve by selecting from the dropdown and clicking the button
  - The groups in the flow json Core.SetGroupMembership block should be updated
  - Edit the text of groups so that the group_key no longer matches the id or the group_name no longer matches the name of the group previously matched in builder.config.json
  - You should see a loading spinner and then you should see the matcher to fix the group you broke
- Test prop mismatch -  `src/store/builder/5b8c87d6-de90-4bc4-8668-4f040000006e-flow-block-group.json`
  - You should see the prop mismatches highlighted
  - You should be able to resolve by selecting from the dropdown and clicking the button
  - The props in the flow json Core.SetContactProperty block should be updated
  - Edit the test of groups so that the property_key no longer matches the name of the property previously matched in builder.config.json
  - You should see a loading spinner and then you should see the matcher to fix the group you broke
- Test a valid flow `src/store/builder/5b8c87d6-de90-4bc4-8668-4f040000006e-flow.json`:
  - You should see the flow rendered correctly
- Test resources work as expected - imported resources correctly associated with blocks
- Test for regressions. 
  - Focus on Create Flow in particular. 
  - Ensure that you can create a blank flow with Create Flow, then import a different flow. 
  - When you return to Home you should see both flows.
  - Test for any other regressions
- Test loop with export panel:
  - Refresh the Home page to remove all flows 
  - Create a new blank flow using the Create Flow page. Add some blocks, configure some resources
  - Copy the JSON from the export panel on the builder page
  - Return to Home and refresh to clear all flows
  - Import the copied JSON
  - The import should be valid and show exactly what you had before the refresh
- Test overwriting existing flows, resources:
  - Flows and resources are identified by the uuid field. 
  - Upload the basic flow example file - `src/store/builder/5b8c87d6-de90-4bc4-8668-4f040000006e-flow.json` and view the flow in the builder. Go to Home and see the flow in the list. 
  - Edit the file to remove some blocks and resources and reupload. 
  - You should see the blocks removed in the builder and the resources removed from associated blocks (e.g. the resource removed from a message block). 
  - Return to Home - you should still see only the one flow. 
- Check validation when no props or groups in builder config
  - Remove the existing props, groups and languages from the builder config file (see the diff with the file on this branch to see where to remove things). You may need to restart `yarn serve` and refresh to load the changes.
  - Attempt the valid flow upload
  - You should see mismatches where you didn't before 
  - It should not be possible to resolve the mismatches as there are no options in the builder config.
- Test that the flow can only contain blocks in `builder.config.json`
  - `src/store/builder/5b8c87d6-de90-4bc4-8668-4f040000006e-flow-unsupported-block` contains unsupported blocks, or you can edit builder.config.json to remove support for block types that are in other flows.
